### PR TITLE
add raincloud plots from Beacon Biosignals

### PR DESCRIPTION
**Pull request by SimonDanisch**
We decided to open source @kimlaberinto amazing raincloud plots and add them to base Makie.
This is a direct port of the docs & the latest implementation.
@kimlaberinto wanted to spent some more time to make the recipe even more usable/pretty before we merge.
We also need to add some tests to the reference image tests.

cc: @haberdashPI @ericphanson @mich11 @palday @jrevels

### CODE REVIEW

**Comment by kimlaberinto** [comment]
@haberdashPI mentioned this in a previous discussion that we've had. I wanted to repost it here so others can see and so that we potentially make this change as well.

>I want to advocate for the API that is made public to be compatible with AlgebraOfGraphics. I think it's a simple change and it would be nice if it worked straightforwardly in that case: basically the interface needs to use a flat array (like columns of a DataFrame) to denote the category labels along x, rather than using nested arrays by default. I was thinking of making a little PR myself to do this...

**Comment by haberdashPI** [comment]
I've added some proposed changes in #1839. See that PR for details. I've tested this with a [PR of AoG](https://github.com/JuliaPlots/AlgebraOfGraphics.jl/pull/383) which supports the latest version of Makie (by just updated the [compat] entries), and that seems to be working well for calling `visual(RainCloud)`.

**Comment by haberdashPI** [comment]
Sorry about that, it looks like my use of some newer keyword syntax is leading to the failures on 1.3: `(;show_median)` instead of `(;show_median=show_median)`

File: src/basic_recipes/raincloud.jl [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    data_array = plot.data_array[]
+    category_labels, data_array = ungroup_labels(category_labels, data_array)
+    if any(ismissing, data_array)
+        error("missing values in data not supported. Please filter out any missing values before plotting")
+    end
+
+    # Checking kwargs, and assigning defaults if they are not in kwargs
+    # General Settings
+    # Define where categories should lie
+    x_positions = if eltype(category_labels) isa String
```
Comment (haberdashPI) [inline src/basic_recipes/raincloud.jl:188]: Whoops, sorry, looks like this needs to change. Sorry, I forgot to run the doctests locally.

```suggestion
    x_positions = if eltype(category_labels) isa AbstractString
```

File: src/basic_recipes/raincloud.jl [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+####
+#### Functions that make the cloud plot
+####
+function plot!(
+        ax::Makie.Axis, P::Type{<: RainClouds},
+        allattrs::Attributes, category_labels, data_array)
+
+    plot = plot!(ax.scene, P, allattrs, category_labels, data_array)
+
+    if eltype(category_labels) <: AbstractString
```
Comment (haberdashPI) [inline src/basic_recipes/raincloud.jl:130]:
```suggestion
    if any(x -> x isa AbstractString, category_labels)
```

File: NEWS.md [hunk 3]
```diff
@@ -5,6 +5,7 @@
 - **Breaking** Added `space` as a generic attribute to switch between data, pixel, relative and clip space for positions. `space` in text has been renamed to `markerspace` because of this. `Pixel` and `SceneSpace` are no longer valid inputs for `space` or `markerspace`.
 - **Breaking** Deprecated `mouse_selection(scene)` for `pick(scene)`.
 - **Breaking** Bumped `GridLayoutBase` version to `v0.7`, which introduced offset layouts. Now, indexing into row 0 doesn't create a new row 1, but a new row 0, so that all previous content positions stay the same. This makes building complex layouts order-independent [#1704](https://github.com/JuliaPlots/Makie.jl/pull/1704).
+- Added `raindclouds` and `raindclouds!` [#1725](https://github.com/JuliaPlots/Makie.jl/pull/1725)
```
Comment (jrevels) [inline NEWS.md:8]: raindclouds -> rainclouds?

**Comment by SimonDanisch** [comment]
Ok, this should build now...
I removed dodge here: https://github.com/JuliaPlots/Makie.jl/pull/1725/commits/ba50702ffe817b34d18cffacb81147641b57f170, since `group_labels` didn't seem to have the ability to deal with dodge...Not sure if zipping dodge in there was just an oversight, or if `group_labels` should be able to deal with that.
Also, it's all gray now, I think before, it was cycling the color?

**Comment by haberdashPI** [comment]
It should be cycling color; I can debug today or tomorrow. Is there anything special I should know to setup the doc build on my machine?

**Comment by SimonDanisch** [comment]
>add raincloud plots from Beacon Biosignals by SimonDanisch · Pull Request #1725 · JuliaPlots/Makie.jl

I think it was doing that before, if I'm not mistaken.. So think something got removed ;)

>Is there anything special I should know to setup the doc build on my machine?

I usually just copy out the code blocks, since the setup is a bit annoying and you can't just build a single page 😅

**Comment by haberdashPI** [comment]
I've made some updates to the docs in #1936 .

This also fixes a few bugs I found in the process.

**Comment by SimonDanisch** [comment]
Test failures due to: https://github.com/JuliaPlots/Makie.jl/pull/1937

File: NEWS.md [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 ## master
 
+- Added `rainclouds` and `raindclouds!` [#1725](https://github.com/JuliaPlots/Makie.jl/pull/1725)
```
Comment (hannahilea) [inline NEWS.md:5]:
```suggestion
- Added `rainclouds` and `rainclouds!` [#1725](https://github.com/JuliaPlots/Makie.jl/pull/1725)
```

File: docs/examples/plotting_functions/rainclouds.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+CairoMakie.activate!(type = "png")
+using Random
+using Makie: rand_localized
+
+####
+#### Below is used for testing the plotting functionality.
+####
+
+function mockup_distribution(N)
+    all_possible_labels = ["Single Mode", "Double Mode", "Single Mode", "Double Mode",
```
Comment (hannahilea) [inline docs/examples/plotting_functions/rainclouds.md:18]: is it intentional that single mode and double mode show up twice here?

File: docs/examples/plotting_functions/rainclouds.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    if category_type == "Single Mode"
+        random_mean = rand_localized(0, 8)
+        random_spread_coef = rand_localized(0.3, 1)
+        data_points = random_spread_coef*randn(N) .+ random_mean
+
+    elseif category_type == "Double Mode"
+        random_mean = rand_localized(0, 8)
+        random_spread_coef = rand_localized(0.3, 1)
+        data_points = random_spread_coef*randn(Int(round(N/2.0))) .+ random_mean
+
```
Comment (hannahilea) [inline docs/examples/plotting_functions/rainclouds.md:31]:
```suggestion
```
potential duplication? otherwise these variables get overridden immediately

**Comment by SimonDanisch** [comment]
Yes, I think that's what's mostly left.
Would also be nice to bring back the color cycling. Let me know if you need help with that.

**Comment by haberdashPI** [comment]
> Would also be nice to bring back the color cycling. Let me know if you need help with that.

I think this is addressed? At least when I ran the help examples on my machine they are colored now. (#1936)

I'll set up a PR to fix @hannahilea's comments.

**Comment by SimonDanisch** [comment]
Ah you're right, great :) Somehow the last time I checked the docs build, it was still all black

File: docs/examples/plotting_functions/rainclouds.md (hunk 6 above)
Comment (haberdashPI) [reply docs/examples/plotting_functions/rainclouds.md:31]: This looks right to me. (Note the `vcat` on line 33).

**Comment by haberdashPI** [comment]
https://github.com/JuliaPlots/Makie.jl/pull/2019 makes a small fix to address @hannahilea's comments.

**Comment by SimonDanisch** [comment]
Yay, thanks everyone :)

**Pull request merged**